### PR TITLE
Treat 'so', 'dylib', 'dll' as executable suffixes

### PR DESCRIPTION
**Pull request by sbc100**
This means we perform linking rather than stopping at bitcode
in these cases.

Also simply the complex error messages we show when using
SIDE_MODULE or MAIN_MODULE when *not* linking.

### CODE REVIEW

File: emcc.py [hunk 1]
```diff
@@ -1624,11 +1623,11 @@ def compile_source_file(i, input_file):
             logger.debug('link: ' + str(linker_inputs) + specified_target)
             # Sort arg tuples and pass the extracted values to link.
             shared.Building.link(linker_inputs, specified_target)
-        logger.debug('stopping at bitcode')
+        logger.debug('stopping at object file')
+        if shared.Settings.SIDE_MODULE:
+          exit_with_error('SIDE_MODULE should only be used when linking')
         if shared.Settings.SIDE_MODULE:
```
Comment (dschuff) [inline emcc.py]: should this line be `if shared.settings.MAIN_MODULE`?

**Comment by brion** [comment]
Would this make it so you would get JS instead of bitcode when building libraries with autoconf + emconfigure + emmake? This would break my workflow for ogv.js, where I build libraries with their native build systems and then link one or more into each JS/wasm module.

Or do I misunderstand?

**Comment by sbc100** [comment]
> Would this make it so you would get JS instead of bitcode when building libraries with autoconf + emconfigure + emmake? This would break my workflow for ogv.js, where I build libraries with their native build systems and then link one or more into each JS/wasm module.
> 
> Or do I misunderstand?

I hope not.  I would assume that any autoconf + emconfigure + emmake setup is going to be using the static library path right?  This should be unaffected.    Or do we try to make these tools think that emscripten does in fact support shared libraries?

**Comment by kripken** [comment]
> do we try to make these tools think that emscripten does in fact support shared libraries?

I believe we do. The reason for that is that in the current model build systems don't build all the way to JS automatically. They build to bitcode, and then the user runs an emcc command on them (usually with some -s flags and an optimization level) to get the final JS. This PR seems to change that as @brion said (`EXECUTABLE_SUFFIXES` will be compiled all the way to JS; I think the test failures here might be because of that).

However, it does seem like we need to change that for the wasm object file case anyhow. So maybe it needs to be rethought.

What is the specific purpose of this current PR, btw? (is it for wasm object files?)

**Comment by brion** [comment]
In a pinch I think I can use the static library .a path if this does change; I just remember having some linking problems with that in the past. (Maybe bad order of operations.)

**Pull request closed without merging**